pkg/kubectl/cmd/util/factory.go

```diff
@@ -344,6 +348,40 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 			}
 			return priorityRESTMapper, api.Scheme
 		},
+		DynamicObject: func() (meta.RESTMapper, runtime.ObjectTyper) {
+			cfg, err := clients.ClientConfigForVersion(nil)
+			CheckErr(err)
+
+			dc, err := discovery.NewDiscoveryClientForConfig(cfg)
+			CheckErr(err)
+
+			groupResources, err := discovery.GetAPIGroupResources(dc)
+			CheckErr(err)
+
+			// Register unknown APIs as third party for now to make
+			// validation happy. TODO perhaps make a dynamic schema
+			// validator to avoid this.
+			for _, group := range groupResources {
+				for _, version := range group.Group.Versions {
+					gv := unversioned.GroupVersion{Group: group.Group.Name, Version: version.Version}
+					if !registered.IsRegisteredVersion(gv) {
+						registered.AddThirdPartyAPIGroupVersions(gv)
+					}
+				}
+			}
+
+			mapper := discovery.NewRESTMapper(groupResources,
+				func(unversioned.GroupVersion) (*meta.VersionInterfaces, error) {
```

Comment:
Since you depend on returning meta you can't put it in runtime (would be a circular reference).  So meta is as close as you can get to runtime.

Also, at some point in the future (probably when we implement the partial object return stuff) we'll want to move unstructured out of runtime and into its own package.